concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+For example, stating that some map has the features to conform to an interface
+only if some entries of some known names and value types are present in the map.
+(However, this is not often seen, so this document will continue to frequently
+refer to "sets of methods" when discussing interfaces.)
+
+### explicit versus structural interfaces
+
+Interfaces as found in programming languages tend to fall into one of two large camps:
+*explicit* interfaces, or *structural* interfaces.
+
+Explicit interfaces require types implementing them to say so.
+
+Structural interfaces state some set of methods (or more generally, features)
+that must be present on a type for it to satisfy the interface;
+then, any type that has those methods implements the interface.
+
+Both explicit and structural styles of interfaces have "open" membership --
+you can "compile" a program that has interfaces in it,
+then add more libraries when you "link" the program and execute it,
+and those libraries can add more implementors of the interfaces!
+
+The explicit and structural styles vary in other practical details.
+For example, explicit interfaces require referencing the interface they implement;
+this makes them significantly less flexible, and creates many dependency management requirements.
+By contrast, structural interfaces can knowingly conform to an interfaces contract
+while still not referencing it directly; this can be extremely useful in
+designing systems for independently upgradeable pieces.
+
+(There are other yet more detailed ways in which these styles vary, but
+since they're not really important in this document, we'll rest here.)
+
+
+Types in the Wild
+-----------------
+
+These different categories of types are present in most programming languages,
+but not all of them are present in every language (!),
+and they often go by different names in different programming communities.
+Here's a few (non-exhaustive) examples in a handful of languages:
+
+### enums in the wild
+
+In Java, the `enum` keyword maps directly the category of [enum](#enums) we've defined here.
+
+In Go, *there is no direct support for enums*.
+However, it's often implemented in practice by a typedef of an int, and a table of named constants.
+
+In Rust, there's an `enum` keyword, and it may be used to implement [enum](#enums) as we've defined it here.
+However, if you're coming from a Rust background, and trying to understand this document and compare to other languages,
+be cautious here: the Rust `enum` keyword *also* provides what we call [sum types](#sum-types).
+
+#### booleans
+
+In many languages, the concept of `boolean` can be understood as a particular instance of an enum --
+it's just an enum with two values!
+Sure, it may be implemented as 'primitive' in many languages... but clearly,
+for cardinality counting, we can see a boolean as exactly "|2|".
+
+### sum types in the wild
+
+In Java, *there is no direct support for sum types*.
```

Comment:
People coming from the JVM world _may_ be familiar with Scala's "sealed traits", which are sum types declared as though they were interface types but with the word `sealed` ensuring external modules can't add new variants.

Reply:
(i.e. a language-supported version of the "create an interface, and promise..." approach)

Reply:
Sum types can be emulated in most languages that have some concept of "private" too tho, with public factory methods, right? You don't really even need the private/public side of this if you have strong language conventions.

Reply:
Instanceof https://github.com/ipld/specs/pull/242#issuecomment-600560619 .

It's true you can produce any of these patterns with enough methods, either compiler-enforced or convention-and-pray, but it becomes increasingly confusing to talk about more ways to do it in this document that just wants to convince people that the concept exists and can be extracted.  I think the Go and Java examples already cover most of the it-isn't-explicitly-supported-so-i-pretended options at a broad level.